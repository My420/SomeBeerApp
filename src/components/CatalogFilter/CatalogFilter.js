import React from 'react';
import styles from './CatalogFilter.module.scss';
import BeerName from './BeerName/BeerName';
import Button from '../Button/Button';
import { PAGE, BEER_NAME_PROP } from '../../utils/constants';
import createOptions from '../../utils/createOptions';
import compareOptions from '../../utils/compareOptions';
import Parameters from './Parameters/Parameters';
import IconSearchSvg from '../IconSearchSvg/IconSearchSvg';

class CatalogFilter extends React.Component {
  constructor(props) {
    super(props);
    const { value } = props;
    this.state = { ...value };
  }

  onPropertyChange = (value, property) => {
    this.setState({ [property]: value });
  };

  onSubmit = evt => {
    evt.preventDefault();
    const { history, value } = this.props;
    if (!compareOptions(this.state, value)) {
      const options = createOptions({ ...this.state, [PAGE]: '1' });
      history.push(options);
    }
  };

  render() {
    // eslint-disable-next-line no-console
    console.log('render ========== CatalogFilter');

    const { [BEER_NAME_PROP]: name } = this.state;

    return (
      <section className={styles.filter}>
        <form className={styles.form} onSubmit={this.onSubmit}>
          <div className={styles.nameWrapper}>
            <BeerName value={name} onPropertyChange={this.onPropertyChange} />
            <Button className={styles.submit} type="submit">
              <IconSearchSvg pathClass={styles.path} />
              <span className="visually-hidden">Apply</span>
            </Button>
          </div>
          <Parameters
            options={this.state}
            onPropertyChange={this.onPropertyChange}
          />
        </form>
      </section>
    );
  }
}

export default CatalogFilter;
