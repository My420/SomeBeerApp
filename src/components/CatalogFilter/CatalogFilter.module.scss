@import '../../sass/variables.scss';

.filter {
  width: 100%;
  padding-bottom: 10px;

  font-size: 18px;
  line-height: 24px;
}

.form {
  width: 350px;
  margin: 0 auto;
}

.nameWrapper {
  display: flex;
  justify-content: space-between;

  width: 100%;

  margin-bottom: 5px;
}

.submit {
  width: 35px;
  height: 35px;

  background-color: transparent;

  border: 2px solid $menu-bar-color;
  border-radius: 5px;
  @include button;
}

@media (max-width: 400px) {
  .form {
    width: 280px;
    margin: 0 auto;
  }
}
