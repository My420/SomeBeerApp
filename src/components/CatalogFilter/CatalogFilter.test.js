import React from 'react';
import { shallow, mount } from 'enzyme';
import CatalogFilter from './CatalogFilter';
import {
  ABV_LESS_PROP,
  ABV_MORE_PROP,
  IBU_LESS_PROP,
  IBU_MORE_PROP,
  EBC_LESS_PROP,
  EBC_MORE_PROP,
  PAGE,
  BEER_NAME_PROP
} from '../../utils/constants';
import createOptions from '../../utils/createOptions';

describe('test <CatalogFilter /> component', () => {
  const value = {
    [BEER_NAME_PROP]: 'test'
  };

  const pushMock = jest.fn();
  const defMock = jest.fn();

  const history = {
    push: pushMock
  };

  const wrapper = shallow(<CatalogFilter value={value} history={history} />);

  beforeEach(() => {
    pushMock.mockClear();
    defMock.mockClear();
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('initialState = props.value', () => {
    expect(wrapper.state()).toEqual(value);
  });

  test('should render Options component with correct props', () => {
    expect(wrapper.find('ToggleWrapper(Parameters)').prop('options')).toEqual(
      value
    );
    expect(
      typeof wrapper.find('ToggleWrapper(Parameters)').prop('onPropertyChange')
    ).toBe('function');
  });

  test('should render BeerName component with correct props', () => {
    expect(wrapper.find('BeerName').prop('value')).toBe(value[BEER_NAME_PROP]);
    expect(typeof wrapper.find('BeerName').prop('onPropertyChange')).toBe(
      'function'
    );
  });

  test('should render Button component with type submit', () => {
    expect(wrapper.find('Button').prop('type')).toBe('submit');
  });

  test('submit event should  call evt.preventDefault', () => {
    wrapper.find('form').simulate('submit', { preventDefault: defMock });
    expect(defMock.mock.calls.length).toBe(1);
  });

  test('submit event should  not call history.push prop if state not changed', () => {
    wrapper.find('form').simulate('submit', { preventDefault: defMock });
    expect(pushMock.mock.calls.length).toBe(0);
  });

  test('submit event should  call history.push prop if state  changed', () => {
    const newState = { [BEER_NAME_PROP]: 'new test' };
    wrapper.setState(newState);
    wrapper.find('form').simulate('submit', { preventDefault: defMock });
    expect(pushMock.mock.calls.length).toBe(1);
  });

  test('history.push shuold call with correct argument if state  changed', () => {
    const newState = { [BEER_NAME_PROP]: 'new test test' };
    const arg = createOptions({ ...newState, [PAGE]: '1' });
    wrapper.setState(newState);
    wrapper.find('form').simulate('submit', { preventDefault: defMock });
    expect(pushMock.mock.calls[0][0]).toBe(arg);
  });
});

describe('test state cahnging <CatalogFilter /> component', () => {
  const value = {
    [ABV_LESS_PROP]: '0',
    [ABV_MORE_PROP]: '0',
    [IBU_LESS_PROP]: '0',
    [IBU_MORE_PROP]: '0',
    [EBC_LESS_PROP]: '0',
    [EBC_MORE_PROP]: '0',
    [PAGE]: '1',
    [BEER_NAME_PROP]: ''
  };

  const pushMock = jest.fn();

  const history = {
    push: pushMock
  };

  const component = mount(<CatalogFilter value={value} history={history} />);

  beforeEach(() => {
    component.setState({ ...value });
    component.find('Button.button').simulate('click');
  });

  test('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  test(`shuld correct change state if input.nameInput change`, () => {
    component
      .find('input.nameInput')
      .simulate('change', { target: { value: 'Test' } });
    const expectedState = { ...value, [BEER_NAME_PROP]: 'Test' };
    expect(component.state()).toEqual(expectedState);
  });

  test(`shuld correct change state if ABV less input change`, () => {
    component
      .find("input[name='abvLess']")
      .simulate('change', { target: { value: '60', name: 'abvLess' } });
    const expectedState = { ...value, [ABV_LESS_PROP]: '60' };
    expect(component.state()).toEqual(expectedState);
  });

  test(`shuld correct change state if ABV more input change`, () => {
    component
      .find("input[name='abvMore']")
      .simulate('change', { target: { value: '70', name: 'abvMore' } });
    const expectedState = { ...value, [ABV_MORE_PROP]: '70' };
    expect(component.state()).toEqual(expectedState);
  });

  test(`shuld correct change state if EBC less input change`, () => {
    component
      .find("input[name='ebcLess']")
      .simulate('change', { target: { value: '10', name: 'ebcLess' } });
    const expectedState = { ...value, [EBC_LESS_PROP]: '10' };
    expect(component.state()).toEqual(expectedState);
  });

  test(`shuld correct change state if EBC more input change`, () => {
    component
      .find("input[name='ebcMore']")
      .simulate('change', { target: { value: '72', name: 'ebcMore' } });
    const expectedState = { ...value, [EBC_MORE_PROP]: '72' };
    expect(component.state()).toEqual(expectedState);
  });

  test(`shuld correct change state if IBU less input change`, () => {
    component
      .find("input[name='ibuLess']")
      .simulate('change', { target: { value: '23', name: 'ibuLess' } });
    const expectedState = { ...value, [IBU_LESS_PROP]: '23' };
    expect(component.state()).toEqual(expectedState);
  });

  test(`shuld correct change state if IBU more input change`, () => {
    component
      .find("input[name='ibuMore']")
      .simulate('change', { target: { value: '12', name: 'ibuMore' } });
    const expectedState = { ...value, [IBU_MORE_PROP]: '12' };
    expect(component.state()).toEqual(expectedState);
  });
});
