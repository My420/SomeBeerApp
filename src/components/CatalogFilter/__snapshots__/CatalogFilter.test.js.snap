// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <CatalogFilter /> component should match snapshot 1`] = `
<section
  className="filter"
>
  <h3
    className="visually-hidden"
  >
    Options
  </h3>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="nameWrapper"
    >
      <BeerName
        onPropertyChange={[Function]}
        value="test"
      />
      <Button
        activeStyle=""
        className="submit"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        <IconSearchSvg
          pathClass="path"
          wrapperClass="container"
        />
        <span
          className="visually-hidden"
        >
          Apply
        </span>
      </Button>
    </div>
    <ToggleWrapper(Parameters)
      onPropertyChange={[Function]}
      options={
        Object {
          "beer_name": "test",
        }
      }
    />
  </form>
</section>
`;

exports[`test state cahnging <CatalogFilter /> component should match snapshot 1`] = `
<CatalogFilter
  history={
    Object {
      "push": [MockFunction],
    }
  }
  value={
    Object {
      "abv_gt": "0",
      "abv_lt": "0",
      "beer_name": "",
      "ebc_gt": "0",
      "ebc_lt": "0",
      "ibu_gt": "0",
      "ibu_lt": "0",
      "page": "1",
    }
  }
>
  <section
    className="filter"
  >
    <h3
      className="visually-hidden"
    >
      Options
    </h3>
    <form
      className="form"
      onSubmit={[Function]}
    >
      <div
        className="nameWrapper"
      >
        <BeerName
          onPropertyChange={[Function]}
          value=""
        >
          <fieldset
            className="nameField"
          >
            <legend
              className="visually-hidden"
            >
              Beer Name
            </legend>
            <input
              className="nameInput"
              maxLength={40}
              onChange={[Function]}
              placeholder="search"
              type="text"
              value=""
            />
          </fieldset>
        </BeerName>
        <Button
          activeStyle=""
          className="submit"
          disabled={false}
          onClick={[Function]}
          type="submit"
        >
          <button
            className="submit "
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
             
            <IconSearchSvg
              pathClass="path"
              wrapperClass="container"
            >
              <span
                className="container"
              >
                <svg
                  height="100%"
                  viewBox="0 0 24 24"
                  width="100%"
                >
                  <path
                    className="path"
                    d="M23.111 20.058l-4.977-4.977c.965-1.52 1.523-3.322 1.523-5.251 0-5.42-4.409-9.83-9.829-9.83-5.42 0-9.828 4.41-9.828 9.83s4.408 9.83 9.829 9.83c1.834 0 3.552-.505 5.022-1.383l5.021 5.021c2.144 2.141 5.384-1.096 3.239-3.24zm-20.064-10.228c0-3.739 3.043-6.782 6.782-6.782s6.782 3.042 6.782 6.782-3.043 6.782-6.782 6.782-6.782-3.043-6.782-6.782zm2.01-1.764c1.984-4.599 8.664-4.066 9.922.749-2.534-2.974-6.993-3.294-9.922-.749z"
                  />
                </svg>
              </span>
            </IconSearchSvg>
            <span
              className="visually-hidden"
            >
              Apply
            </span>
             
          </button>
        </Button>
      </div>
      <ToggleWrapper(Parameters)
        onPropertyChange={[Function]}
        options={
          Object {
            "abv_gt": "0",
            "abv_lt": "0",
            "beer_name": "",
            "ebc_gt": "0",
            "ebc_lt": "0",
            "ibu_gt": "0",
            "ibu_lt": "0",
            "page": "1",
          }
        }
      >
        <Parameters
          isOpen={true}
          onPropertyChange={[Function]}
          options={
            Object {
              "abv_gt": "0",
              "abv_lt": "0",
              "beer_name": "",
              "ebc_gt": "0",
              "ebc_lt": "0",
              "ibu_gt": "0",
              "ibu_lt": "0",
              "page": "1",
            }
          }
          toggleState={[Function]}
        >
          <Button
            activeStyle=""
            className="button"
            disabled={false}
            onClick={[Function]}
          >
            <button
              className="button "
              disabled={false}
              onClick={[Function]}
              type="button"
            >
               
              <span
                className="buttonText"
              >
                hide
              </span>
               
            </button>
          </Button>
          <TransitionGroup
            childFactory={[Function]}
            component={null}
          >
            <CSSTransition
              classNames={Object {}}
              in={true}
              key=".0"
              onExited={[Function]}
              timeout={300}
            >
              <Transition
                appear={false}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={false}
              >
                <div>
                  <InputFields
                    onPropertyChange={[Function]}
                    options={
                      Object {
                        "abv_gt": "0",
                        "abv_lt": "0",
                        "beer_name": "",
                        "ebc_gt": "0",
                        "ebc_lt": "0",
                        "ibu_gt": "0",
                        "ibu_lt": "0",
                        "page": "1",
                      }
                    }
                  >
                    <div
                      className="wrapper"
                    >
                      <Abv
                        lessValue="0"
                        moreValue="0"
                        onPropertyChange={[Function]}
                      >
                        <fieldset
                          className="abvField"
                        >
                          <legend
                            className="abvLegend"
                          >
                            Alcohol By Volume
                          </legend>
                          <div
                            className="abvInner"
                          >
                            <label
                              className="abvLabelMore"
                              htmlFor="abvMore"
                            >
                              <span
                                className="textMore"
                              >
                                more than
                              </span>
                              <input
                                className="abvInputMore"
                                id="abvMore"
                                name="abvMore"
                                onChange={[Function]}
                                type="number"
                                value="0"
                              />
                            </label>
                            <label
                              className="abvLabelLess"
                              htmlFor="abvLess"
                            >
                              <span
                                className="textLess"
                              >
                                less than
                              </span>
                              <input
                                className="abvInputLess"
                                id="abvLess"
                                name="abvLess"
                                onChange={[Function]}
                                type="number"
                                value="0"
                              />
                            </label>
                          </div>
                        </fieldset>
                      </Abv>
                      <Ibu
                        lessValue="0"
                        moreValue="0"
                        onPropertyChange={[Function]}
                      >
                        <fieldset
                          className="ibuField"
                        >
                          <legend
                            className="ibuLegend"
                          >
                            International Bitterness Unit
                          </legend>
                          <div
                            className="ibuInner"
                          >
                            <label
                              className="ibuLabelMore"
                              htmlFor="ibuMore"
                            >
                              <span
                                className="textMore"
                              >
                                more than
                              </span>
                              <input
                                className="ibuInputMore"
                                id="ibuMore"
                                name="ibuMore"
                                onChange={[Function]}
                                type="number"
                                value="0"
                              />
                            </label>
                            <label
                              className="ibuLabelLess"
                              htmlFor="ibuLess"
                            >
                              <span
                                className="textLess"
                              >
                                less than
                              </span>
                              <input
                                className="ibuInputLess"
                                id="ibuLess"
                                name="ibuLess"
                                onChange={[Function]}
                                type="number"
                                value="0"
                              />
                            </label>
                          </div>
                        </fieldset>
                      </Ibu>
                      <Ebc
                        lessValue="0"
                        moreValue="0"
                        onPropertyChange={[Function]}
                      >
                        <fieldset
                          className="ebcField"
                        >
                          <legend
                            className="ebcLegend"
                          >
                            Color Units Ebc
                          </legend>
                          <div
                            className="ebcInner"
                          >
                            <label
                              className="ebcLabelMore"
                              htmlFor="ebcMore"
                            >
                              <span
                                className="textMore"
                              >
                                more than
                              </span>
                              <input
                                className="ebcInputMore"
                                id="ebcMore"
                                name="ebcMore"
                                onChange={[Function]}
                                type="number"
                                value="0"
                              />
                            </label>
                            <label
                              className="ebcLabelLess"
                              htmlFor="ebcLess"
                            >
                              <span
                                className="textLess"
                              >
                                less than
                              </span>
                              <input
                                className="ebcInputLess"
                                id="ebcLess"
                                name="ebcLess"
                                onChange={[Function]}
                                type="number"
                                value="0"
                              />
                            </label>
                          </div>
                        </fieldset>
                      </Ebc>
                    </div>
                  </InputFields>
                </div>
              </Transition>
            </CSSTransition>
          </TransitionGroup>
        </Parameters>
      </ToggleWrapper(Parameters)>
    </form>
  </section>
</CatalogFilter>
`;
